### DISCUSSION TITLE
Reorganize repo to make it easier to work with

### DISCUSSION THREAD

**Discussion by gempain** [outline/outline#1269] [General]
Hi,

First, I'd like to say that using Outline is awesome. The UI is beautiful and I love the fact that it's really simple to use. A while back, I opened #1177 to push towards enhancing deployment experience and authentication.

As #1183 is making progress towards a better way to integrate PassportJS authentication methods (@jakoblorz has moved forward quite a bit with the PR, and I've written the Mattermost integration [here](https://github.com/jakoblorz/outline/pull/1)), I feel that the next step is to move the structure of the repository to something more developer friendly and productivity-driven. I know that you're focusing on your paid solution, and it makes sense, but you would benefit a lot from a repo that is friendly to newcomers that want to help adding features or fixing bugs.

I've been playing around with this repository, and I would have several suggestions for making it easier to work with. I don't have the time to push these suggestions forward in a PR, and I feel that it's not the community's role to do so, you are best suited as you know the code best.

## Develop outside of Docker
It's nice to have everything build inside the Dockefile, but in practice, it's just super slow, the npm install takes forever and the server is slow in dev. I'm on a 16GB 2015 MBP, latest OS X, SSD drive, 4 x 2.3Ghz CPU with 8 threads, yet it takes at least 3-5 minutes just to start developing when you first install the app and at least 10-15s to authenticate. Docker for Mac isn't so good at synchronizing files with the file system, and developing is a nightmare in these conditions.

What I would suggest is to only include redis, as3 and postgres in the `docker-compose.yml` and `npm run dev` locally - I mean, everyone has Node installed locally, and I personally don't see the benefit of developing in a Docker container.

## Move server code into an `src` directory and webpack server code

Ideally, you'd want separate git repositories for the frontend and backend part of Outline, but at least it would be really nice to split files into a clear hierarchy. 
- `src/server`
- `src/ui`

The build script should create:
- `dist/server`: webpacked server code, with a bit of effort it would even be possible to webpack the `node_modules` and make the bundle quite small, hence making the final docker image very light and fast to build
- `dist/ui`: built React app, just like now, except in a subdirectory `ui`

## Don't hard code too much in the UI, particularly the URL

The UI should be a static app, which can be configured with an environment file (e.g. `environment.json`). This way, you don't always have to re-build the UI whenever you change the URL.

## Split the UI and server into 2 docker images

You already have a nicely designed API, why not just making the frontend and backend completely separate ? The UI could use the `create-react-app` skeleton. Separation of concerns if the key.

## Cleanup repo files

There are A LOT of directories and files in the root directory. Could this be simplified ?
```
.circleci # keep this
.githooks # necessary ?
.github # necessary ?
.vscode # this is idea specific, IMO shouldn't be in the repo

app # -> src/
server # -> src/
public # -> src/
shared # -> src/

CODE_OF_CONDUCT.md # could be merged with README ?
SECURITY.md # could be merged with README ?
Procfile # essential ?
.eslintignore # needed ? could probably be moved inside `package.json`
flow-typed # can be removed if moving to TS
init.js # used by flow (can be removed if moving to TS) or Heroku (AFAIK there's no Outline demo on Heroku, so why is this here ? If it's to help people deploy on heroku, then I guess its fine)
app.json # used by Heroku if I'm not mistaken.  If it's to help people deploy on heroku, then I guess its fine
Makefile # developing outside of docker => no need for this file - just explain in readme what to do
.babelrc # could be moved inside `package.json` ? fine otherwise, I understand the separation of concern
.sequelizerc # could be moved inside `package.json` ? fine otherwise, I understand the separation of concern
.eslintrc # could be moved inside `package.json` ? fine otherwise, I understand the separation of concern
.flowconfig # no need if moving to TS
webpack.config.dev.js # could get away with a single webpack file
webpack.config.js # could get away with a single webpack file
webpack.config.prod.js # could get away with a single webpack file
```

## Move to Typescript

IMHO, Typescript has become very mature, and moving from JS/Flow to Typescript isn't really complicated. I think the repo would really benefit from it.

---

I hope these suggestions are helpful and that you don't take them as criticizm. It's an honest attempt to help make this repo the greatest knowledge base on GH. And for what it's worth, it's not because it's easy to setup, develop and deploy that you will lose paid customers. The community will love you for it.

Keep up the great work !

**Comment by tommoor** [comment]
Hey, thanks for writing all this up, it's a lot! For the most part I agree with you, and I want to be clear that in no way have we tried to obfuscate development or the running Outline to try and push people towards the hosted product. I'll give some quick answers to your points so you know where I'm at:

#### Develop outside of Docker

I'm down for this change, I've personally felt the pain of yarn install inside docker more than anyone else. The docker-compose setup is there for the OSS community, If this was a private repo It would be much easier to just run the services locally 😉 

####  Ideally, you'd want separate git repositories for the frontend and backend part of Outline

I disagree with this, by most standards this is a tiny monorepo and the benefits to productivity far outweigh any learning curve from having server and ui in the same repo.

####  it would be really nice to split files into a clear hierarchy.

The files are already in `/app`, `/server`, and `/shared` – this is documented in the readme. We could move inside a `/src` directory but that doesn't seem like a big deal either way. I agree that `ui` would be clearer than `app`!

#### Split the UI and server into 2 docker images

The current idea is that you can run the entire application on a single node server – to keep things simple, so this one seems like a potential dev improvement at the cost of deployment complexity?

#### Don't hard code too much in the UI, particularly the URL

Do you have an example of this? All of the ui variables should be resolved during the build process from ENV variables.

#### There are A LOT of directories and files in the root directory

I'm sure a couple of these can be removed but really it's not a lot for a product of this complexity. I don't see the benefit of nesting everything in deep directory structures, that makes things less accessible. Similarly having the config in dot files in the root is very standard and most of the MD files are part of the way GitHub works.

#### Move to Typescript

I'd love to do this, I just moved the editor over to Typescript: https://github.com/outline/rich-markdown-editor and I think it'll happen this year. Like most of the above, it's just a matter of time – If I take several weeks moving the codebase to typescript that's weeks I'm not working on features and bugs, so balance is key.

**Reply by gempain** [reply]
Hey @tommoor, thanks for your answer, I'm glad you took it the right way 😄 I'm convinced that in no way you've tried to obfuscate development/deployment, no worries about that.

> Hey, thanks for writing all this up, it's a lot! For the most part I agree with you, and I want to be clear that in no way have we tried to obfuscate development or the running Outline to try and push people towards the hosted product. I'll give some quick answers to your points so you know where I'm at:

No worries, I am convinced it's not obsfucation. I know that any repo has history and that there's always room for improvement.

> #### Develop outside of Docker
> I'm down for this change, I've personally felt the pain of yarn install inside docker more than anyone else. The docker-compose setup is there for the OSS community, If this was a private repo It would be much easier to just run the services locally 😉

Sweet. As I said, redis, as3 and postgres should stay in the docker-compose, not in a million years do you want to deploy those locally without Docker. The typical way IMO of building a good docker image is to compile things locally (most times in your CI) and then copy the `dist` folder in your Docker image. Copying is quick, but any processing inside that image will kill performance. Once you have a clean image, you can publish it on Dockerhub and people don't need to do anything beyond running it and configuring env variables in compose or via `docker run -e`.

The problem for me is that the current folder hierarchy doesn't allow building a Docker image easily. I tried, but I gave up: as is, I have to create a `.dockerignore` and start whitelisting files to `COPY` in the image, and it's a trial/error process as I'm not sure what most files are for. If you had everything in an `src` directory, it'd be easy to determine what needs to go in the image. If we compiled everything with webpack, it would be really nice as well because you would just have to copy the `dist` folder in your Docker image, and that's it. After that, people don't even need to clone your repo anymore to deploy, they just copy the docker compose, `up` and that's it. You can't do more simple than that 😄 

> #### Ideally, you'd want separate git repositories for the frontend and backend part of Outline
> I disagree with this, by most standards this is a tiny monorepo and the benefits to productivity far outweigh any learning curve from having server and ui in the same repo.

Fair enough 😄 

> #### it would be really nice to split files into a clear hierarchy.
> The files are already in `/app`, `/server`, and `/shared` – this is documented in the readme. We could move inside a `/src` directory but that doesn't seem like a big deal either way. I agree that `ui` would be clearer than `app`!

Yup, I recognized the `app` directory because usually React apps are placed in `src/app`, but TBH I wasn't certain.

> #### Split the UI and server into 2 docker images
> The current idea is that you can run the entire application on a single node server – to keep things simple, so this one seems like a potential dev improvement at the cost of deployment complexity?

Fair enough.

> #### Don't hard code too much in the UI, particularly the URL
> Do you have an example of this? All of the ui variables should be resolved during the build process from ENV variables.

Well, I've seen that you use the webpack define plugin to hardcode those variables at compile time. Usually, what we do is to only hardcode things that will never change after the build: version number, app name, conditional compilation stuff, etc. Everything that can be configured, like API URL, is just a simple `environment.json` placed in the public directory, and then in the UI you create a react provider like so:

```jsx
export interface Environment {
  production: boolean;
  debug: boolean;
  name: string;
  api: {
    url: string;
    timeout: number;
  };
}

export const EnvContext = createContext<Environment>(undefined);

export const useEnv = () => useContext(EnvContext);

export function EnvProvider(props) {
  const [loading, setLoading] = useState(true);
  const [env, setEnv] = useState<Environment>();
  const [error, setError] = useState();

  useEffect(() => {
    fetch('/assets/environments/environment.json')
      .then(res => res.json())
      .then(loadedEnv => {
        setEnv(loadedEnv);
        console.log('environment', loadedEnv);
      })
      .finally(() => setLoading(false))
      .catch(setError);
  }, []);

  if (loading) {
    return <FullPageLoader/>
  }

  if (error) {
    return <FullPageCentered>
      <AlertError error={error}/>
    </FullPageCentered>
  }

  return <EnvContext.Provider value={env} {...props}/>;
}
```

then you just place that around your entire app, and in any component you can do
```jsx
const env = useEnv();
```

Correct me if I'm wrong, but with the current implementation, if I publish port 3002 of my docker container, the UI won't be able to reach the backend as it was built with 3000.

> #### There are A LOT of directories and files in the root directory
> I'm sure a couple of these can be removed but really it's not a lot for a product of this complexity. I don't see the benefit of nesting everything in deep directory structures, that makes things less accessible. Similarly having the config in dot files in the root is very standard and most of the MD files are part of the way GitHub works.

I know that CODE_OF_CONDUCT and SECURITY files are the way GH does it, but honestly I'm not a fan of their way 😛 We end up having so much noise in the repos that you get lost in all the files. I know some people like dot files, I do too, but I always try to only keep the essential. Anyways, cleaning up what's not needed would help, even just a couple of files.

> #### Move to Typescript
> I'd love to do this, I just moved the editor over to Typescript: https://github.com/outline/rich-markdown-editor and I think it'll happen this year. Like most of the above, it's just a matter of time – If I take several weeks moving the codebase to typescript that's weeks I'm not working on features and bugs, so balance is key.

Honestly, your (and our) life would be much more simple that way. You could do this very easily:
1. Delete everything flow related, cleanup flow annotations
2. Add a basic tsconfig (make sure you have `strict: false` enabled in the beginning)
3. Rename all files from `js` to `ts`

That's it really, it should compile right away. Then, it's up to you to bundle with webpack or not, but honestly, you'd go from having to copy 250Mb of `node_modules` to having a single bundle of 1.5Mb... think about it 😄 By moving to TS, you'll need an `src` directory, it'll simplify a lot.

**Comment by jakoblorz** [comment]
> #### Develop outside of Docker
> It's nice to have everything build inside the Dockefile, but in practice, it's just super slow, the npm install takes forever and the server is slow in dev. I'm on a 16GB 2015 MBP, latest OS X, SSD drive, 4 x 2.3Ghz CPU with 8 threads, yet it takes at least 3-5 minutes just to start developing when you first install the app and at least 10-15s to authenticate. Docker for Mac isn't so good at synchronizing files with the file system, and developing is a nightmare in these conditions.
> What I would suggest is to only include redis, as3 and postgres in the docker-compose.yml and npm run dev locally - I mean, everyone has Node installed locally, and I personally don't see the benefit of developing in a Docker container.

You can always run only a subset of the services defined in the `docker-compose.yml` with `docker-compose up [-d] s3 redis postgres`. 

I personally (XPS 13 2019) have set `DOCKER_HOST`, so that the containers are run on a more powerful machine. Using port-forwarding (`ssh -L localhost:5432:<DOCKER_HOST IP>:5434 -L localhost:4569:<DOCKER_HOST IP>:4569 -L localhost:6379:<DOCKER_HOST IP>:6380 username@<DOCKER_HOST IP>`) I am then able to only run outline on my laptop, given a correctly configured .dev.env

> #### Move to Typescript
> I'd love to do this, I just moved the editor over to Typescript: https://github.com/outline/rich-markdown-editor and I think it'll happen this year. Like most of the above, it's just a matter of time – If I take several weeks moving the codebase to typescript that's weeks I'm not working on features and bugs, so balance is key.

> Honestly, your (and our) life would be much more simple that way. You could do this very easily:

> Delete everything flow related, cleanup flow annotations
Add a basic tsconfig (make sure you have strict: false enabled in the beginning)
Rename all files from js to ts
That's it really, it should compile right away. Then, it's up to you to bundle with webpack or not, but honestly, you'd go from having to copy 250Mb of node_modules to having a single bundle of 1.5Mb... think about it smile By moving to TS, you'll need an src directory, it'll simplify a lot.

Correct me if I am wrong, but I think there will always be a different compilation / transpilation pipeline for server and app. I am a fan of TypeScript as well, but I don't think that it is that easy. In fact, this [repo](https://github.com/niieani/typescript-vs-flowtype) tracks the subtle or not-so-subtle differences between flow and typescript in syntax.

**Reply by gempain** [reply]
Yes, there will still be two compilation flows. A single webpack file can be used but with multiple configs.

As for TS, I'm not talking about the learning curve (which IMO isn't that big 😄, surprising for a Microsoft product 😛) but definitely easy to switch to technically. I'l give it a shot some evening that I have time to spare 😄
